argh/tests/lib.rs

```diff
@@ -94,6 +94,455 @@ fn subcommand_example() {
+  one               First subcommand.
+  two               Second subcommand.
+"###;
+
+    match TopLevel::from_args(&["cmdname"], &["--help"]) {
+        Ok(_) => panic!("help was parsed as args"),
+        Err(e) => {
+            assert_eq!(help_str, e.output);
+            e.status.expect("help returned an error");
+        }
+    }
+}
+
+#[test]
+fn optional_subcommand_empty_show_help_for_both() {
+    #[derive(FromArgs, PartialEq, Debug)]
+    /// Top-level command.
+    struct TopLevel {
+        #[argh(subcommand, default = "one")]
+        nested: Option<MySubCommandEnum>,
+    }
+
+    #[derive(FromArgs, PartialEq, Debug)]
+    #[argh(subcommand)]
+    enum MySubCommandEnum {
+        One(SubCommandOne),
+        Two(SubCommandTwo),
+    }
+
+    #[derive(FromArgs, PartialEq, Debug)]
+    /// First subcommand.
+    #[argh(subcommand, name = "one")]
+    struct SubCommandOne {
+        #[argh(option)]
+        /// how many x
+        x: usize,
+    }
+
+    #[derive(FromArgs, PartialEq, Debug)]
+    /// Second subcommand.
+    #[argh(subcommand, name = "two")]
+    struct SubCommandTwo {
+        #[argh(switch)]
+        /// whether to fooey
+        fooey: bool,
+    }
+
+    let help_str = r###"Usage: cmdname [<command>] [<args>]
+
+Top-level command.
+
+Options:
+  --help            display usage information
+
+Commands:
+  one               First subcommand.
+  two               Second subcommand.
+
+When no subcommand is given, the command `one` is used as the default. Allowing the additional:
+
+First subcommand.
+
+Options:
+  --x               how many x
+  --help            display usage information
+"###;
+
+    match TopLevel::from_args(&["cmdname"], &["--help"]) {
+        Ok(_) => panic!("help was parsed as args"),
+        Err(e) => {
+            assert_eq!(help_str, e.output);
+            e.status.expect("help returned an error");
+        }
+    }
+}
+
+#[test]
+fn option_subcommand_empty() {
+    #[derive(FromArgs, PartialEq, Debug)]
+    /// Top-level command.
+    struct TopLevel {
+        #[argh(subcommand)]
+        nested: Option<MySubCommandEnum>,
+    }
+
+    #[derive(FromArgs, PartialEq, Debug)]
+    #[argh(subcommand)]
+    enum MySubCommandEnum {
+        One(SubCommandOne),
+        Two(SubCommandTwo),
+    }
+
+    #[derive(FromArgs, PartialEq, Debug)]
+    /// First subcommand.
+    #[argh(subcommand, name = "one")]
+    struct SubCommandOne {
+        #[argh(option)]
+        /// how many x
+        x: usize,
+    }
+
+    #[derive(FromArgs, PartialEq, Debug)]
+    /// Second subcommand.
+    #[argh(subcommand, name = "two")]
+    struct SubCommandTwo {
+        #[argh(switch)]
+        /// whether to fooey
+        fooey: bool,
+    }
+
+    let one = TopLevel::from_args(&["cmdname"], &[]).expect("sc 1");
+    assert_eq!(one, TopLevel { nested: None },);
+}
+
+#[test]
+fn option_subcommand_with_default() {
+    #[derive(FromArgs, PartialEq, Debug)]
+    /// Top-level command.
+    struct TopLevel {
```

Comment:
what happens if we have a name collision between top-level fields, and the default subcommand field? For example, say we change this struct to:

```rust
struct TopLevel {
    #[argh(option)]
    /// how many toplevel x
    x: String,

    nested: Option<MySubCommandEnum>,
}
```

we'd collide on `--x`. I'm guessing the TopLevel field would probably take precedence. Ideally this would be an error, but I don't think we have enough info to enforce this in argh_derive.